lib/ddtrace/writer.rb

```diff
@@ -29,53 +31,63 @@ def initialize(options = {})
 
       @traces_flushed = 0
 
-      # spawns two different workers for spans and services;
-      # they share the same transport which is thread-safe
-      @trace_worker = nil
-      @service_worker = nil
+      # one worker for both services and traces, each have their own queues
+      @worker = nil
     end
 
     # spawns two different workers for spans and services;
     # they share the same transport which is thread-safe
     def start
-      @trace_worker = Datadog::Workers::AsyncTransport.new(@span_interval, @transport, @buff_size) do |items, transport|
-        send_spans(items, transport)
-      end
-      @service_worker = Datadog::Workers::AsyncTransport.new(@service_interval, @transport, 1) do |items, transport|
-        send_services(items, transport)
+      @next_send_services = Time.now
+      @worker = Datadog::Workers::AsyncTransport.new(@span_interval,
+                                                     @transport,
+                                                     [:traces, :services],
+                                                     @buff_size) do |items, transport|
+        send(items, transport)
       end
 
-      @trace_worker.start()
-      @service_worker.start()
+      @worker.start()
     end
 
     # stops both workers for spans and services.
     def stop
-      @trace_worker.stop()
-      @trace_worker = nil
-      @service_worker.stop()
-      @service_worker = nil
+      @worker.stop()
+      @worker = nil
+    end
+
+    # flush spans to the trace-agent, handles both spans & services
+    def send(items, transport)
+      return if items.empty?
+      if items[0].instance_of? Array
+        send_spans(items, transport)
+      elsif items[0].instance_of? Hash
+        if Time.now >= @next_send_services
+          send_services(items, transport)
+          @next_send_services = Time.now + @service_interval
+        else
+          @worker.enqueue(:services, items)
+        end
+      end
     end
 
-    # flush spans to the trace-agent
+    # flush spans to the trace-agent, handles spans only
     def send_spans(traces, transport)
       # FIXME[matt]: if there's an error, requeue; the new Transport can
       # behave differently if it's a server or a client error. Don't requeue
       # if we have a client error?
       spans = Datadog::Encoding.encode_spans(traces)
       transport.send(SPANS_ENDPOINT, spans)
 
-      # TODO[all]: is it really required? it's a number that will grow indefinitely
       @traces_flushed += traces.length()
     end
 
-    # flush services to the trace-agent
+    # flush services to the trace-agent, handles services only
```

Comment:
I know that I did that, but I prefer to remove this behavior. I don't like this "automatic" rescheduling when the service doesn't change.

I will update our ``set_service_info`` in the ``tracer.rb`` so that:
* if someone calls that, we will store the value
* if the value is already there AND is the same, don't bother sending the same stuff again
* if the tracer is enabled and the writer is set, send that value to the ``writer``

In the ``writer.rb``:
* if our method is called, enqueue the new ``service``
* the ``AsyncTransport`` will fire the call when the time comes.

Later we should think how the ``AsyncTransport`` can handle the failure (i.e. the agent is not started at the beginning). But I think this is a great improvement that makes the flow easier and more reliable.

Does it make sense?